oauth_dropins/indieauth.py

```diff
@@ -0,0 +1,128 @@
+"""IndieAuth drop-in.
+
+https://indieauth.com/developers
+"""
+
+import json
+import logging
+import urllib
+import urlparse
+
+import appengine_config
+import handlers
+import models
+from webutil import util
+
+import bs4
+import requests
+from google.appengine.ext import ndb
+
+
+INDIEAUTH_URL = 'https://indieauth.com/auth'
+
+
+def discover_authorization_endpoint(me):
+  """Fetch a URL and look for authorization_endpoint Link header or
+  rel-value.
+
+  Args:
+    me: string, URL to fetch
+
+  Return:
+    string, the discovered indieauth URL or the default indieauth.com URL
+  """
+  logging.debug('trying to fetch user url "%s"', me)
+  resp = util.requests_get(me)
+  if resp.status_code < 200 or resp.status_code >= 300:
+    logging.debug(
+      'could not fetch user url "%s". got response code: %d',
+      me, resp.status_code)
+    return INDIEAUTH_URL
+  # check authorization_endpoint header first
+  auth_endpoint = resp.links.get('authorization_endpoint', {}).get('url')
+  if auth_endpoint:
+    return auth_endpoint
+  # check the html content
+  soup = bs4.BeautifulSoup(resp.text)
+  auth_link = soup.find('link', {'rel': 'authorization_endpoint'})
+  auth_endpoint = auth_link and auth_link['href']
```

Comment:
forgot BS's API...will this raise `KeyError` if `auth_link` doesn't have `href`?